# Methylation Analysis

## dms

- **Description**: Get the differentially methylated sites between two samples.

- **Usage**: `cgmaptools dms [-i <CGmapInter>] [-m 5 -M 100] [-o output]`

    ```
    -i FILE               File name for CGmapInter, STDIN if omitted
    -m INT, --min=INT     min coverage [default : 0]
    -M INT, --max=INT     max coverage [default : 100]
    -o OUTFILE            To standard output if omitted. Compressed output if
                          end with .gz
    -t STRING, --test-method=STRING
                          chisq, fisher [default : chisq]
    ```

- **Example**:

    \#1 Using the output of `intersect` as input:
    
    ```
    Chr1  C  3541  CG  CG  0.8  4  5  0.4  4  10
    ```
    
    The output of `dms` is:
    
    ```
    chr1	C	4654	CG	CG	0.92	1.00	8.40e-01
    chr1	C	4658	CHH	CC	0.50	0.00	3.68e-04
    chr1	G	8376	CG	CG	0.62	0.64	9.35e-01
    ```
    
## dmr

- **Description**: Get the differentially methylated region by Fisher's exact test.

- **Usage**: `cgmaptools dmr [-i <CGmapInter>] [-m 5 -M 100] [-o output]`

    ```
    -i FILE                 File name for CGmapInter, STDIN if omitted
    -c INT, --minCov=INT    min coverage [default : 0]
    -C INT, --maxCov=INT    max coverage [default : 100]
    -s INT, --minStep=INT   min step in bp [default : 100]
    -S INT, --maxStep=INT   max step in bp [default : 500]
    -n INT, --minNSite=INT  min N sites [default : 2]
    -o OUTFILE              To standard output if omitted. Compressed output if
                            end with .gz
    ```
    
- **Example**:

    \#1 Using the output of `intersect` as input:
    
    ```
    Chr1  C  3541  CG  CG  0.8  4  5  0.4  4  10
    ```
    
    The output of `dms` is:
    
    ```
    chr1	1004572	1004574	inf	    0.00e+00	0.1100	0.0000
    chr1	1009552	1009566	-0.2774	8.08e-01	0.0200	0.0300
    chr1	1063405	1063498	0.1435	8.93e-01	0.6333	0.5733
    ```
    
## asm

- **Description**: Allele specific methylation analysis.

- **Usage**: `cgmaptools asm [options] -r <ref.fa> -b <input.bam> -l <snp.vcf>`

    ```
    -r    Samtools indexed reference genome seqeunce, fasta format. eg. hg19.fa
          - use samtools to index reference first: samtools faidx hg19.fa
    -b    Samtools indexed Bam format file.
          - use samtools to index bam file first: samtools index <input.bam>
    -l    SNPs in vcf file format.
    -s    Path to samtools eg. /home/user/bin/samtools
          - by defualt, we try to search samtools in your system PATH,
    -o    Output results to file. [default: STDOUT]
    -t    C context. [default: CG]
          - available context: C, CG, CH, CW, CC, CA, CT, CHG, CHH
    -m    Specify calling mode. [default: asr]
          - alternative: ass
          - asr: allele specific methylated region
          - ass: allele specific methylated site
    -d    Minimum number of read for each allele linked site to call ass. [default: 3]
          - ass specific.
    -n    Minimum number of C site each allele linked to call asr. [default: 2]
          - asr specific.
    -D    Minimum read depth for C site to call methylation level when calling asr.
          [default: 1]
          - asr specific.
    -L    Low methylation level threshold. [default: 0.2]
          - allele linked region [or site] with low methylation level should be 
            no greater than this threshold.
    -H    High methylation level threshold. [default: 0.8]
          - allele linked region[or site] with high methylation level should be
            no less than this threshold.
    -q    Adjusted p value using Benjamini & Hochberg (1995) 
          ("BH" or its alias "fdr"). [default: 0.05]
    ```
  
## mbed

- **Description**: Calculate average methylation levels in given regions.

- **Usage**: `cgmaptools mbed [-i <CGmap>]  -b <regin.bed> [-c 5 -C 500 -s]`

    ```
    -i  String, CGmap file; use STDIN if not specified
        Ex: chr1	G	3000851	CHH	CC	0.1	1	10
    -b  String, BED file
        Ex: chr1	3000000	3005000	-
    -c  Int, minimum Coverage [Default: 5]
    -C  Int, maximum Coverage [Default: 500]
    -s  Strands would be distinguished when specified
    ```
    
- **Example**:

    The output format:
    
    ```
    
    ```

## mbin

- **Description**: Generate the methylation level in Bins.

- **Usage**: `cgmaptools mbin [-i <CGmap>] [-c 10 --CXY 5 -B 5000000]`

    ```
    -i FILE               File name end with .CGmap or .CGmap.gz. 
                          If not specified, STDIN will be used.
    -B BIN_SIZE           Define the size of bins [Default: 5000000]
    -c COVERAGE           The minimum coverage for site selection [Default: 10]
    -C CONTEXT, --context=CONTEXT
                          specific context: CG, CH, CHG, CHH, CA, CC, CT, CW
                          use all sites if not specified
    --cXY=COVERAGEXY      Coverage for chrX/Y should be half that of autosome
                          for male [Default: same with -c]
    -f FIGTYPE, --figure-type=FIGTYPE
                          png, pdf, eps. Will not generate figure if not
                          specified
    -p STRING             Prefix for output figures
    -t STRING, --title=STRING
                          title in the output figures
    ```

- **Example**:

    The output format:
    
    ```
    chr1    1       5000    0.0000
    chr1    5001    10000   0.0396
    chr2    1       5000    0.0755
    chr2    5001    10000   0.0027
    chr3    1       5000    na
    ```

## mmbin

- **Description**: Generate the methylation level in Bins for multiple samples.

- **Usage**: `cgmaptools mmbin [-l <1.CGmap[,2.CGmap,..]>] [-c 10 --CXY 5 -B 5000000]`

    ```
    -l FILE               File name list, end with .CGmap or .CGmap.gz. If not
                        specified, STDIN will be used.
    -t FILE               List of samples
    -B BIN_SIZE           Define the size of bins [Default: 5000000]
    -C CONTEXT, --context=CONTEXT
                          specific context: CG, CH, CHG, CHH, CA, CC, CT, CW
                          use all sites if not specified
    -c COVERAGE           The minimum coverage for site selection [Default: 10]
    --cXY=COVERAGEXY      Coverage for chrX/Y should be half that of autosome
                          for male [Default: same with -c]
    ```

- **Example**:

    The output format:
    
    ```
    chr1    1       5000    0.0000
    chr1    5001    10000   0.0396
    chr2    1       5000    0.0755
    chr2    5001    10000   0.0027
    chr3    1       5000    na
    ```

## mfg

- **Description**: Calculated methylation profile across fragmented regions.

- **Usage**: `cgmaptools mfg [-i <CGmap>]  -r <region> [-c 5 -C 500]`

    ```
    -i  String, CGmap file; use STDIN if not specified
        chr1	G	851	CHH	CC	0.1	1	10
    -r  String, Region file, at least 4 columns
        Format: chr	strand	pos_1	pos_2	pos_3	...
        Regions would be considered as [pos_1, pos_2), [pos_2, pos_3)
        Strand information will be used for distinguish sense/antisense strand
        Ex:
        chr1	+	600	700	800	900	950
        chr1	-	1600	1500	1400	1300	1250
    -c  Int, minimum Coverage [Default: 5]
    -C  Int, maximum Coverage [Default: 500]
        Sites exceed the coverage range will be discarded
    ```

- **Example**:

    The output format:
    
    ```
    Region_ID       R_1     R_2     R_3     R_4
    sense_ave_mC    0.50    0.40    0.30    0.20
    sense_sum_mC    5.0     4.0     3.0     2.0
    sense_sum_NO    10      10      10      10
    anti_ave_mC     0.40    0.20    0.10    NaN
    anti_sum_mC     8.0     4.0     2.0     0.0
    anti_sum_NO     20      20      20      0
    total_ave_mC    0.43    0.27    0.17    0.2
    total_sum_mC    13.0    8.0     5.0     2.0
    total_sum_NO    30      30      30      10
    ```

## mstat

- **Description**: Methyaltion statistic.

- **Usage**: `cgmaptools mstat [-i <CGmap>]`

    ```
    -i FILE               File name end with .CGmap or .CGmap.gz. If not
                          specified, STDIN will be used.
    -c COVERAGE           The minimum coverage for site selection [Default: 10]
    -f FILE, --figure-type=FILE
                          png, pdf, eps. Will not generate figure if not
                          specified
    -p STRING             Prefix for output figures
    -t STRING, --title=STRING
                          title in the output figures
    ```

- **Example**:

    The output format:
    
    ```
    MethStat        context C       CG      CHG	    CHH	    CA      CC      CT      CH      CW
    mean_mC         global  0.0798  0.3719  0.0465  0.0403  0.0891  0.0071  0.0241  0.0419  0.0559
    sd_mCbyChr      global  0.0078  0.0341  0.0163  0.0110  0.0252  0.0049  0.0076  0.0096  0.0148
    count_C         global  10000   1147    2332    6521    3090    2539    3224    8853    6314
    contrib_mC      global  1.0000  0.5348  0.1360  0.3292  0.3452  0.0228  0.0973  0.4652  0.4424
    quant_mC        [0]     8266    471     2012    5783    2422    2421    2952    7795    5374
    quant_mC   (0.00 ,0.20] 705     182     155     368     272     97      154     523     426
    mean_mC_byChr   chr1    0.0840  0.4181  0.0340  0.0412  0.0794  0.0065  0.0251  0.0393  0.0513
    mean_mC_byChr   chr10   0.0917  0.4106  0.0758  0.0421  0.0968  0.0097  0.0349  0.0502  0.0655
    ```

## mtr

- **Description**: Calculate the methylation levels in regions in two ways.

- **Usage**: `cgmaptools mtr [-i <CGmap>] -r <region> [-o <output>]`

    ```
    -i FILE     File name end with .CGmap or .CGmap.gz. If not specified, STDIN
                will be used.
    -r FILE     Filename for region file, support *.gz
    -o OUTFILE  To standard output if not specified.
    ```

- **Example**:

    The input file format:
    
    ```
    #chr    start_pos  end_pos
    chr1    8275       8429
    ```

    The output format:
    
    ```
    #chr  start_pos  end_pos  mean(mC)  #_C  #read(C)/#read(T+C)  #read(T+C)
    chr1   8275       8429     0.34     72         0.40             164
    ```
    