# Differential Methylation

For bisulfite sequencing projects, differential methylation calling is a regular procedure to have a better understanding of sample identities. In most cases, the differences between treatment (normal) and control (disease) are rare at DNA methylome level. The challenges to find these differences with high accuracy lie in high sensitivity of tools used as well as enough coverage and depth of sequencing data.

To compare the methylomes of different samples, we defined **bin** along the chromosomes with dynamic number of cytosines covered by both samples. Only bins meet the following conditions were used for comparisons:

1. The minimum number of cytosine in a bin is 5.
2. The maximum distance between adjacent cytosine covered by both samples should be 100bp.
3. The maximum size of a bin should be 1000bp.

Bins were determined as DMRs if the p value obtained from t test less than 0.01 and the average methylation level between two samples greater than 10%. All these threshoulds were adjustable to fit the actual situations.

At last, DMRs with distance less than 1000bp were merged together.

```{r, echo=FALSE, out.width="70%", fig.align='center', fig.cap="Schematic diagram of DMR callling"}
knitr::include_graphics(path = "./DMR_snapshot.png")
```

```
Usage: CGmapInterDiffReg.py [-i <CGmapInter>] [-m 5 -M 100] [-o output]

Description:
  Get the differentially methylated regions by Fisher's exact test.
Input Format, same as the output of CGmapIntersect.py:
   Chr1  C  3541  CG  CG  0.8  4  5  0.4  4  10
Output Format, Ex:
   chr1	1004572	1004574	inf	    0.00e+00	0.1100	0.0000
   chr1	1009552	1009566	-0.2774	8.08e-01	0.0200	0.0300
   chr1	1063405	1063498	0.1435	8.93e-01	0.6333	0.5733
   
Options:
  -h, --help            show this help message and exit
  -i FILE               File name for CGmapInter, STDIN if omitted
  -c INT, --minCov=INT  min coverage [default : 0]
  -C INT, --maxCov=INT  max coverage [default : 100]
  -s INT, --minStep=INT
                        min step in bp [default : 100]
  -S INT, --maxStep=INT
                        max step in bp [default : 500]
  -n INT, --minNSite=INT
                        min N sites [default : 2]
  -o OUTFILE            To standard output if omitted. Compressed output if
                        end with .gz
```



