# Visualization

## Lollipop plot

**Description**: Plot the site-specific methylation levels for multiple samples.

**Usage**: `mCviewer.R [options] file`

**Options**:

```
	-i INFILE, --infile=INFILE
		input file

	-a ANNOTATION, --annotation=ANNOTATION
		[opt] sample name

	-o OUTFILE, --outfile=OUTFILE
		[opt] output file

	-f FORMAT, --format=FORMAT
		[opt] the format for output figure: pdf (default), png, eps

	-l LEFT, --left=LEFT
		[opt] Left-most position

	-r RIGHT, --right=RIGHT
		[opt] Right-most position

	-c CHR, --chr=CHR
		[opt] chromosome name

	-t TITLE, --title=TITLE
		[opt] text shown on title

	-w WIDTH, --width=WIDTH
		[opt] width (in inch). Default: 8.

	--height=HEIGHT
		[opt] height (in inch). Default: 8.

	-s SITE, --site=SITE
		[opt] file of site to be marked

	-b BED, --bed=BED
		[opt] BED file for region to be markered

	-h, --help
		Show this help message and exit
```

**Example**:

```
  ../bin/mCviewer.R [-i input] -o gene.png
-Input Format:
    >= 3 columns, 1st line is the header, using R color name or "NaN".
    Can be output by CGmapFillIndex.py. Use STDIN if omitted.
  Example:
    chr  pos  E_vs_EMT	EMT_vs_M	E_vs_M
    chr1	13116801	NaN	NaN	darkgreen
    chr1	13116899	NaN	red	NaN
-BED File Format:
    the first 4 columns are required
  Example:
    chr1  213941196  213942363  REGION-1
    chr1  213942363  213943530  REGION-2
```

**Example (figure)**:

```{r, echo=FALSE, out.width="80%", fig.align='center', fig.cap="Lollipop plot showing DNA methylation"}
knitr::include_graphics(path = "./Snai1.png")
```


## Tanghulu plot

**Description**: Representing the methylation status of mapped reads to Tanghulu plot.

**Usage**: `CircleGraphics.pl [options] -r <ref.fa> -b <input.bam> -l <region eg. chr1:133333-134444 | position eg. chr1:133333>`

**Options**:

```
  
  -r              Samtools indexed reference genome seqeunce, fasta format. eg. hg19.fa
                  eg. using samtools to index reference: samtools faidx <hg19.fa|hg19.fa.gz>
                  both gz compressed or uncompressed format are supported.
  
  -b              Samtools indexed Bam file to view.
                  eg. using samtools to index bam file: samtools index <input.bam>
  
  -l              Region in which to display DNA methylation.
                  Or specify a single position (eg. heterozygous SNP site), we will show allele specific methylation.
  
  -s              Path to samtools eg. /home/user/bin/samtools
                  By defualt, we try to search samtools in your system,
                  and errors occur if we can not find it.
                  Better to set the path yourself.
  
  -o              Output results to specified file instead of [default: CirclePlot.Ctype.region.Date.pdf].
  
  -t              C context. [default: CG]
                  available context: C, CG, CH, CW, CC, CA, CT, CHG, CHH
  
  -d              Ouput device. [default: pdf]
                  alternative: png
  
  -c              Seperate reads by chain. [default: OFF]
                  specify this option to turn ON.
  
  -v              Show vague allele linked reads. [ default: OFF]
  
  -g              Genotype of heterozygous SNP site.
                  * This option provides two alleles of htSNP site. eg. AT
                  * The genotype information can be used to reduce vague alleles.
                  * This option is specific to display methylation in allele specific mode.
  
  -D              Minimum number of reads (depth) covered in this region or allele linked. [default: 0|OFF]
  
  -C              Minimum number of C (specified type) covered in this region or allele linked. [default: 0|OFF]
  
  -W              Width of graphics reigon in inches. [default: 4]
  
  -H              Height of graphics reigon in inches. [default: 4]
  
  -R              Resolution in ppi. [default: 300]
                  Only available for png device.
  
  -h              Help message.
```

**Example (figure)**:

```{r, echo=FALSE, out.width="80%", out.height="500px", fig.align='center', fig.cap="Tanghulu plot showing DNA methylation of CpG sites on each read"}
knitr::include_graphics(path = "./CirclePlot.CG.chr1_3017150_3017200.2016-12-05.png")
```


## Coverage along chromosomes

**Description**: Generate the overall coverage in Bins.

**Usage**: `ATCGmapCovInBins.py [-i <ATCGmap>] [-B 5000000]`

**Options**:

```
  -h, --help            show this help message and exit
  -i FILE               File name end with .ATCGmap or .ATCGmap.gz. If not
                        specified, STDIN will be used.
  -B BIN_SIZE           Define the size of bins [Default: 5000000]
  -f FILE, --figure-type=FILE
                        png, pdf, eps. Will not generate figure if not
                        specified
  -p STRING             Prefix for output figures
```

**Output Example**:

```
   chr1    1       5000    29.0000
   chr1    5001    10000   30.0396
   chr2    1       5000    35.0755
   chr2    5001    10000   40.0027
   chr3    1       5000    na
```

**Example (figure)**:

```{r, echo=FALSE, out.width="80%", fig.align='center', fig.cap="Reads coverage along chromosomes (CGmap)"}
knitr::include_graphics(path = "./coverageAlongChromosomes_CGmap.png")
```
```{r, echo=FALSE, out.width="80%", fig.align='center', fig.cap="Reads coverage along chromosomes (ATCGmap)"}
knitr::include_graphics(path = "./coverageAlongChromosomes_ATCGmap.png")
```

## Methylation along chromosomes

**Usage**: `CGmapMethylInFragmentedRegion [-i <CGmap>]  -r <region> [-c 5 -C 500 ]`

**Options**:

```
   -i  String, CGmap file; use STDIN if not specified
       chr1	G	851	CHH	CC	0.1	1	10
   -r  String, Region file, at least 4 columns
       Format: chr	strand	pos_1	pos_2	pos_3	...
       Regions would be considered as [pos_1, pos_2), [pos_2, pos_3)
       Strand information will be used for distinguish sense/antisense strand
       Ex:
       chr1	+	600	700	800	900	950
       chr1	-	1600	1500	1400	1300	1250
   -c  Int, minimum Coverage [Default: 5]
   -C  Int, maximum Coverage [Default: 500]
       Sites exceed the coverage range will be discarded
   -h  help
```

**Output Example**:

```
   Region_ID       R_1     R_2     R_3     R_4
   sense_ave_mC    0.50    0.40    0.30    0.20
   sense_sum_mC    5.0     4.0     3.0     2.0
   sense_sum_NO    10      10      10      10
   anti_ave_mC     0.40    0.20    0.10    NaN
   anti_sum_mC     8.0     4.0     2.0     0.0
   anti_sum_NO     20      20      20      0
   total_ave_mC    0.43    0.27    0.17    0.2
   total_sum_mC    13.0    8.0     5.0     2.0
   total_sum_NO    30      30      30      10
```

**Example (figure)**:

```{r, echo=FALSE, out.width="80%", fig.align='center', fig.cap="DNA methylation level along chromosomes (CpG)"}
knitr::include_graphics(path = "./DNAmethylationAlongChr_CG.png")
```
```{r, echo=FALSE, out.width="80%", fig.align='center', fig.cap="DNA methylation level along chromosomes (CH)"}
knitr::include_graphics(path = "./DNAmethylationAlongChr_CH.png")
```

## Methylation in fragmented regions

**Description**: Plot methylation dynamics of target and flanking region for multiple samples.

**Usage**: `mCFragRegView.R [options]`

**Options**:

```
	-i INFILE, --infile=INFILE
		input file

	-r RATIO, --ratio=RATIO
		[opt] range ratio between target region and flanking region in plot

	-o OUTFILE, --outfile=OUTFILE
		[opt] output file name

	-W WIDTH, --width=WIDTH
		[opt] width of figure (inch)

	-H HEIGHT, --height=HEIGHT
		[opt] height of figure (inch)

	-f FORMAT, --format=FORMAT
		[opt] format of output figure. Alternative: png

	-R RESOLUTION, --resolution=RESOLUTION
		[opt] Resolution in ppi. Only available for png format.

	-h, --help
		Show this help message and exit
```


**Example**:

```
  FragRegView.R -i input -r 3 -o genebody.pdf
  -Input File Format:
  1st line is the header.
  Each row contains methylation measurements of a sample.
  Example:
  Sample  Up1  Up2  ...  Region1  Region2 ...  Down1  Down2  ...
  Sample1 0.1  0.1  ...  0.2      0.2     ...  0.3    0.3    ...
  Sample2 0.1  0.1  ...  0.2      0.2     ...  0.3    0.3    ...
```

**Example (figure)**:

```{r, echo=FALSE, out.width="80%", fig.align='center', fig.cap="DNA methylation in target and flanking regions"}
knitr::include_graphics(path = "./genebody.png")
```

## Methylation heatmap

**Description**: Plot methylation dynamics of target region for multiple samples in heatmap.

**Usage**: `../bin/mCBinHeatmap.R [options]`

**Options**:

```
	-i INFILE, --infile=INFILE
		input file

	-o OUTFILE, --outfile=OUTFILE
		[opt] output file name

	-l COLORLOW, --colorLow=COLORLOW
		[opt] color used for the lowest methylation value.

	-m COLORMID, --colorMid=COLORMID
		[opt] color used for the middle methylation value.

	-g COLORHIGH, --colorHigh=COLORHIGH
		[opt] color used for the highest methylation value.

	-n COLORNUMBER, --colorNumber=COLORNUMBER
		[opt] desired number of color elements in the panel.

	-W WIDTH, --width=WIDTH
		[opt] width of figure (inch)

	-H HEIGHT, --height=HEIGHT
		[opt] height of figure (inch)

	-f FORMAT, --format=FORMAT
		[opt] format of output figure. Alternative: png

	-R RESOLUTION, --resolution=RESOLUTION
		[opt] Resolution in ppi. Only available for png format.

	-h, --help
		Show this help message and exit
```

**Example**:

```
  mCBinHeatmap.R -i input -m white -o chr1.xxx-xxx.pdf
  -Input File Format:
  1st line is the header.
  Each column contains methylation measurements of a sample.
  Example:
  Region  Sample1  Sample2 ...
  Region1 0.1      0.1     ...
  Region2 0.1      0.1     ...
```

**Example (figure)**:

```{r, echo=FALSE, out.width="80%", fig.align='center', fig.cap="Heatmap representing DNA methylation in target region"}
knitr::include_graphics(path = "./chr1.methylationHeatmap.png")
```

## Methylation by context

**Description**: Generate the DNA methylation level by context.

**Usage**: `CGmapStatMeth.py [-i <CGmap>]`

**Options**:

```
  -h, --help            show this help message and exit
  -i FILE               File name end with .CGmap or .CGmap.gz. If not
                        specified, STDIN will be used.
  -f FILE, --figure-type=FILE
                        png, pdf, eps. Will not generate figure if not
                        specified
  -p STRING             Prefix for output figures
```

**Output Example**:

```
   MethStat        context C       CG      CHG	    CHH	    CA      CC      CT      CH      CW
   mean_mC         global  0.0798  0.3719  0.0465  0.0403  0.0891  0.0071  0.0241  0.0419  0.0559
   sd_mCbyChr      global  0.0078  0.0341  0.0163  0.0110  0.0252  0.0049  0.0076  0.0096  0.0148
   count_C         global  10000   1147    2332    6521    3090    2539    3224    8853    6314
   contrib_mC      global  1.0000  0.5348  0.1360  0.3292  0.3452  0.0228  0.0973  0.4652  0.4424
   quant_mC        [0]	    8266    471     2012    5783	2422	2421	2952	7795	5374
   quant_mC   (0.00 ,0.20]	705     182     155     368     272     97      154     523     426
   mean_mC_byChr   chr1    0.0840  0.4181  0.0340  0.0412  0.0794  0.0065  0.0251  0.0393  0.0513
   mean_mC_byChr   chr10   0.0917  0.4106  0.0758  0.0421  0.0968  0.0097  0.0349  0.0502  0.0655
```

**Example (figure)**:

```{r, echo=FALSE, out.width="80%", fig.align='center', fig.cap="Average DNA methylation by context"}
knitr::include_graphics(path = "./DNAmethylationLevel_ByContext.png")
```
```{r, echo=FALSE, out.width="80%", fig.align='center', fig.cap="Distribution of DNA methylation by context"}
knitr::include_graphics(path = "./DNAmethylationDistribution_ByContext.png")
```
